### ISSUE TITLE
Minimum node version in engines field makes hnvm inoperable

### ISSUE THREAD

**Issue by TrevorBurnham** [UrbanCompass/hnvm#11]
From a directory with a `package.json` that has an `engines` field with a range, hnvm attempts to use the entire expression as if it were a version number. Example:

```jsonc
// package.json
{
  "engines": {
    "node": ">=14.0.0"
  }
}
```

```console
$ node -v
Downloading node v>=14.0.0 to /Users/trevor.burnham/.hnvm/node
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 10858    0 10858    0     0   100k      0 --:--:-- --:--:-- --:--:--  100k
tar: Error opening archive: Unrecognized archive format
```

**Comment by joseph-galindo** [comment]
We looked at this recently and came across a couple different problems that make this something we are unlikely to fix soon.

Whenever a semver range is given instead of an exact, we intend to take the version range and query against npmjs' https apis to get an exact version. Once we have an exact version, it's trivial to download everything we need from whichever registries are provided.

There are a few problems with things the way they are:
- We'd need to change https://github.com/UrbanCompass/hnvm/blob/82a29a2c70c6e816a9a785b35dfbb07ae782a417/lib/hnvm/config.sh#L123 to `=~ ^[0-9]+\.[0-9]+\.[0-9]+$`. Right now it effectively will only catch semver ranges like `"latest"`, and NOT match on ranges like `">=1.0.0"`
- The curl call at https://github.com/UrbanCompass/hnvm/blob/82a29a2c70c6e816a9a785b35dfbb07ae782a417/lib/hnvm/config.sh#L132 has to change. Right now it hits the HTTP api, which will just 301 redirect to the HTTPS api. Since curl isn't using the -L flag curl will not follow the 301 along to the new URL, so these calls will always fail (well, at least not give any useful response). The curl here needs to change to target https directly, or use -L to follow through to the https URL.
- If we were to fix the above two issues, we'd still have problems at the API level itself. Their API docs for the endpoint we use are here: https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md#getpackageversion . From testing I've done, this endpoint only supports EXACT version numbers, or specifically the "latest" semver range. What this means is that calls like `curl "https://registry.npmjs.org/react-chunk/7.3.0" are fine, but calls like `curl "https://registry.npmjs.org/react-chunk/>=7"` will instead return `{"code":"MethodNotAllowedError","message":"GET is not allowed"}`.

So for the moment, we have some more fundamental API issues to address before we can move further on this.

**Comment by patrickhulce** [comment]
fix is up in https://github.com/UrbanCompass/hnvm/pull/23 :)

**Comment by TrevorBurnham** [comment]
I'm still seeing this issue. When running in the [modus-sdk](https://github.com/modustech/modus-sdk) project, which has `engines` set to `"node": ">=16.0.0"`, I get this from hnvm v0.14.0:

```cli
modus-sdk $ npm install
Resolved node >=16.0.0 to >=16.0.0
```

There's no further output, but the process exits with an error code, and no installation occurs.

If I change the `engines` field to `"node": "16.14.2"`, then the install works as expected:

```cli
modus-sdk $ npm install
Resolved node 16.14.2 to 16.14.2
Using Hermetic NodeJS v16.14.2
...
added 831 packages, and audited 832 packages in 7s
```

**Comment by patrickhulce** [comment]
heh, the fix only works if you don't put a full version in the range 😆 

https://github.com/UrbanCompass/hnvm/blob/cdebf2cea2801d9eead352cecdb1e270adaf49c3/lib/hnvm/config.sh#L127-L129

need to tighten up this regex with some `^$`
